src/tests/profiler/native/gcheapenumerationprofiler/gcheapenumerationprofiler.cpp

```diff
@@ -0,0 +1,237 @@
+        eventMask |= COR_PRF_MONITOR_GC;
+    }
+
+    if (FAILED(hr = pCorProfilerInfo->SetEventMask2(eventMask, COR_PRF_HIGH_MONITOR_NONE)))
+    {
+        _failures++;
+        printf("FAIL: ICorProfilerInfo::SetEventMask2() failed hr=0x%x", hr);
+    }
+
+    return hr;
+}
+
+HRESULT STDMETHODCALLTYPE GCHeapEnumerationProfiler::GarbageCollectionStarted(int cGenerations, BOOL generationCollected[], COR_PRF_GC_REASON reason)
+{
+    SHUTDOWNGUARD();
+    printf("GCHeapEnumerationProfiler::GarbageCollectionStarted\nSleeping for 10 seconds\n");
+    std::this_thread::sleep_for(std::chrono::seconds(10));
+    return S_OK;
+}
+
+HRESULT STDMETHODCALLTYPE GCHeapEnumerationProfiler::GarbageCollectionFinished()
+{
+    SHUTDOWNGUARD();
+    printf("GCHeapEnumerationProfiler::GarbageCollectionFinished\n");
+    return S_OK;
+}
+
+HRESULT GCHeapEnumerationProfiler::Shutdown()
+{
+    Profiler::Shutdown();
+
+    if (_expectedExceptions == 1)
+    {
+        printf("GCHeapEnumerationProfiler::Shutdown: PASS: Encountered exception as expected.\n");
+        printf("PROFILER TEST PASSES\n");
+        fflush(stdout);
+        return S_OK;
+    }
+
+    if (_objectsCount < 100)
+    {
+        printf("GCHeapEnumerationProfiler::Shutdown: FAIL: Expected at least 100 objects, got %d\n", _objectsCount.load());
+        _failures++;
+    }
+
+    if (_customGCHeapObjectTypesCount != 1)
+    {
+        printf("GCHeapEnumerationProfiler::Shutdown: FAIL: Expected 1 custom GCHeapObject type, got %d\n", _customGCHeapObjectTypesCount.load());
+        _failures++;
+    }
+
+    if (_failures == 0)
+    {
+        printf("PROFILER TEST PASSES\n");
+    }
+    else
+    {
+        // failures were printed earlier when _failures was incremented
+    }
+    fflush(stdout);
+
+    return S_OK;
+}
+
+String GCHeapEnumerationProfiler::GetClassIDNameHelper(ClassID classId) {
+    return GetClassIDName(classId);
+}
+
+struct CallbackState
+{
+    std::atomic<int>* objectsCount;
+    GCHeapEnumerationProfiler *instance;
+    std::atomic<int>* customGcHeapObjectTypeCount;
+};
+
+static BOOL STDMETHODCALLTYPE heap_walk_fn(ObjectID object, void* callbackState)
+{
+    CallbackState* state = static_cast<CallbackState*>(callbackState);
+
+    state->objectsCount->fetch_add(1, std::memory_order_relaxed);
+
+    ClassID classId{0};
+    HRESULT hr = state->instance->pCorProfilerInfo->GetClassFromObject(object, &classId);
+    if (hr != S_OK)
+    {
+        printf("Error: failed to get class ID from object.\n");
+        // Returning FALSE will stop the enumeration, just skip this object and continue
+        return TRUE;
+    }
+
+    String classIdName = state->instance->GetClassIDNameHelper(classId);
+    String expectedCustomObjectClass = reinterpret_cast<const WCHAR*>(u"CustomGCHeapObject");
+    if (classIdName == expectedCustomObjectClass)
+    {
+        state->customGcHeapObjectTypeCount->fetch_add(1, std::memory_order_relaxed);
+    }
+
```

Comment:
Lets verify that you can't re-entrantly call EnumerateGCHeapObjects from inside the object callback. I'd expect CORPROF_E_UNSUPPORTED_CALL_SEQUENCE.

Reply:
Like in https://github.com/dotnet/runtime/pull/103735/files#r1671330380, its hitting a contract violation first from calling a GC_TRIGGERS within a GC_NOTRIGGER (from the callback EEToProfInterfaceImpl::EnumerateGCHeapObjectsCallback)